### PR TITLE
Index documents from published_documents message exchange

### PR DESCRIPTION
**Pull request by bishboria**
Currently, changes to topics and other tags in content-tagger don't cause the search index (accessed via rummager) to be updated correspondingly. We want such changes to be reflected as soon as possible in the search index.

Part of: https://trello.com/c/lRnxI2x5/472-make-rummager-index-the-tags-in-the-links-hash

### CODE REVIEW

File: lib/tasks/message_queue.rake [hunk 1]
```diff
@@ -0,0 +1,19 @@
+namespace :message_queue do
+  desc "Index documents that are published to the publishing-api"
+  task :index_documents_from_publishing_api do
+    ENV["RABBITMQ_HOSTS"] = "localhost"
```
Comment (rboulton) [inline lib/tasks/message_queue.rake]: I think these should be set in puppet, rather than in the app; puppet is the place that we normally put configuration.

Also, this value for RABBITMQ_HOSTS won't be right in production.

File: lib/index_documents.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    topic_content_ids = topic_content_ids(message)
+
+    unless topic_content_ids.nil? # no topics field in links, don't update
+      topic_base_paths = topic_content_ids.map { |content_id| base_path(content_id) }
+      update_index(base_path, topic_base_paths)
+    end
+
+    message.ack
+  rescue Exception => e # bury any errors, retry
+    message.retry
```
Comment (rboulton) [inline lib/index_documents.rb]: Why bury the errors?  We should at least send them to errbit.

Comment (benlovell) [inline lib/index_documents.rb]: Also, it's generally better to rescue from at the least `StandardError` (thus its derivatives) so you're not rescuing from runtime issues.

File: config/rabbitmq.yml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+defaults: &defaults
+  host: localhost
+  port: 5672
+  vhost: /
+  user: publishing_api
+  pass: publishing_api
+  recover_from_connection_close: true
```
Comment (rboulton) [inline config/rabbitmq.yml]: Missing newline at the end of this file.

File: config/rabbitmq.yml [hunk 4]
```diff
@@ -0,0 +1,7 @@
+defaults: &defaults
```
Comment (rboulton) [inline config/rabbitmq.yml]: What uses this config file?  We're passing the configuration in via environment variables, aren't we?

File: lib/tasks/message_queue.rake [hunk 5]
```diff
@@ -0,0 +1,21 @@
+namespace :message_queue do
+  desc "Index documents that are published to the publishing-api"
+  task :index_documents_from_publishing_api do
+    # The following environment variables need to be set
+    # RABBITMQ_HOSTS
+    # RABBITMQ_PORT
```
Comment (rboulton) [inline lib/tasks/message_queue.rake]: RABBITMQ_PORT doesn't appear to be a thing that exists in the message consumer code.  I think we're just using the default AMQP port always.

File: lib/index_documents.rb [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+  end
+
+  private
+
+  def document_base_path(message)
+    message.payload["base_path"]
+  end
+
+  def topic_content_ids(message)
+    message.payload["links"].try(:[], "topics")
```
Comment (rboulton) [inline lib/index_documents.rb]: I'm not familiar with this idiom, and am having difficulty working out what it does.  Is it equivalent to `message.payload["links"].fetch("topics", {})` for the purposes of what we're doing?

Comment (benlovell) [inline lib/index_documents.rb]: That would fail if `payload["links"]` is `nil`.

File: lib/index_documents.rb [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+class IndexDocuments
+  def process(message)
+    base_path = document_base_path(message)
+    topic_content_ids = topic_content_ids(message)
+
+    unless topic_content_ids.nil? # no topics field in links, don't update
+      topic_base_paths = topic_content_ids.map { |content_id| base_path(content_id) }
+      update_index(base_path, topic_base_paths)
+    end
```
Comment (rboulton) [inline lib/index_documents.rb]: We need to do this for all the tag types that can be be tagged in content-tagger.  So: mainstream_browse_pages, topics, and organisations (at present).

File: lib/index_documents.rb [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+  def base_path(content_id)
+    GdsApi::PublishingApiV2.new(Plek.current.find('publishing-api')).get_content(content_id)["base_path"]
+  end
+
+  def update_index(base_path, topic_base_paths)
+    current_index.amend(base_path, { topics: topic_base_paths })
+  end
+
+  def current_index
+    @current_index ||= search_server.index(settings.default_index_name)
```
Comment (rboulton) [inline lib/index_documents.rb]: This won't work if the documents aren't in the `mainstream` index, I think.

File: config/rabbitmq.yml (hunk 4 above)
Comment (bishboria) [inline config/rabbitmq.yml]: I added this when starting the ticket, I'll remove the file

File: lib/index_documents.rb (hunk 8 above)
Comment (rboulton) [inline lib/index_documents.rb]: I think the best approach would be to use `search_server.index_for_search(settings.search_config.content_index_names)` to first to a "get document", to check if the document exists at all and find out which index it's in.  Then use search_server.index(indexname) for the index returned in the data for that document to update the document in the right index.

This will also avoid errors being thrown by the call to `amend` if/when we get updates for documents which are excluded from the search index (for whatever reason).

File: lib/index_documents.rb (hunk 6 above)
Comment (rboulton) [inline lib/index_documents.rb]: Ah.  From the documentation, it looks like that will then return `nil` instead of `[]`, which will then fail when the return value is passed to `map`.

Would `message.payload.fetch("links", {}).fetch("topics", [])` be better?

Comment (bishboria) [inline lib/index_documents.rb]: `topic_content_ids(message)` can validly return nil, depending if `links` hash is present, or if there is no `topics` field.

This seemed like the nicest way to avoid `nil["topics"]` at runtime

File: lib/index_documents.rb (hunk 7 above)
Comment (bishboria) [inline lib/index_documents.rb]: This was what I wanted to discuss with you: missing trello ticket information.

Should the ticket be updated and then add extra code changes to this PR?

File: lib/index_documents.rb (hunk 6 above)
Comment (benlovell) [inline lib/index_documents.rb]: Shame we're not using ruby 2.3 and `#dig`

File: lib/index_documents.rb (hunk 8 above)
Comment (bishboria) [inline lib/index_documents.rb]: I'll go with that, you know `rummager` a lot better :)

File: lib/index_documents.rb (hunk 6 above)
Comment (rboulton) [inline lib/index_documents.rb]: With the code as it stands, if `topic_content_ids()` returns `nil`, the call to `topic_content_ids.map` in IndexDocuments::process() will raise a `NoMethodException`.

Comment (rboulton) [inline lib/index_documents.rb]: Ah, scratch that - you're explicitly checking for nil.

File: lib/index_documents.rb [hunk 9]
```diff
@@ -0,0 +1,65 @@
+require_relative "../app"
+require 'gds_api/publishing_api_v2'
+
+class IndexDocuments
+  def process(message)
+    links = message.payload["links"].try(:dup)
+
+    if links
+      links.each do |(k, vs)|
```
Comment (rboulton) [inline lib/index_documents.rb]: Could you use multi-word names for `k` and `vs` here to make it easier to read the code?  Perhaps `v` is `link_type` and `vs` is `content_ids`?

File: lib/index_documents.rb [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+        paths = vs.map { |v| base_path(v) }.compact
```
Comment (rboulton) [inline lib/index_documents.rb]: It would be good to sort the output for this, since the ordering we receive them in is not significant.  Sorting will ensure we get a consistent order out when we process the same document, and that in turn will ensure we don't send unnecessary changes to elasticsearch, allowing it to avoid doing work when there's no semantic change to the document.

File: lib/index_documents.rb [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+        links[k] = paths
+      end
+
+      base_path = document_base_path(message)
+      update_index(base_path, links)
+    end
+
+    message.ack
+  rescue StandardError => e
```
Comment (rboulton) [inline lib/index_documents.rb]: What would happen if we didn't do this rescue at all?

I think that errors will then be caught by the message queue consumer gem that called this method, the consumer will exit (and report the error) and then upstart will restart it and the message will be reprocessed.  This would be functionally very similar to this code, but would also trigger alerting if a message was failing too frequently (because the process would restart frequently, too).  It would also mean less code to test and maintain.

File: lib/index_documents.rb [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    indices_with_base_path(base_path).map do |index|
+      index.amend(base_path, links)
+    end
+  end
+
+  def indices_with_base_path(document_base_path)
+    indices_to_search = search_server.index_for_search(search_config.content_index_names)
+    results = indices_to_search.raw_search(query:{term:{link: document_base_path}})
+
+    indices = results["hits"]["hits"].map do |h|
```
Comment (rboulton) [inline lib/index_documents.rb]: Perhaps rename `h` to `hit`?

File: lib/index_documents.rb [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+    Airbrake.notify_or_ignore(
+      error: e,
+      payload: message.payload,
+      headers: message.headers,
+      delivery_info: message.delivery_info
+    )
+    message.retry
+  end
+
+  private
```
Comment (rboulton) [inline lib/index_documents.rb]: Our style guide says to outdent private (ie, don't have leading space for it here).  Also that it shouldn't have a blank line following it.

File: lib/index_documents.rb (hunk 9 above)
Comment (rboulton) [inline lib/index_documents.rb]: We don't want to use all the link types - only the ones which come from content-tagger.  Also, the values for link types don't always match the field names in rummager (eg, the link type of `topics` maps to the `specialist_sectors` field in rummager), so I think we'll need some custom logic for each link type.

File: test/integration/index_documents_test.rb [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+
+  def teardown
+    clean_test_indexes
+  end
+
+  def test_topics_get_indexed_from_publishing_api
+    m = GovukMessageQueueConsumer::MockMessage.new(payload_with_topics)
+    IndexDocuments.new.process(m)
+
+    assert m.acked?
```
Comment (rboulton) [inline test/integration/index_documents_test.rb]: I think we need to be checking more than that the method acks the message.  Perhaps check that the appropriate payload was sent to elasticsearch?  Or that the topics from the payload were sent to elasticsearch, at least.

File: test/support/index_document_test_helpers.rb [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+    stub_request(:post, "http://localhost:9200/mainstream_test/_bulk").
+      with(:body => post_to_elasticsearch).
+      to_return(:status => 200, :body => elasticsearch_post_response)
+
+    stub_request(:post, "http://localhost:9200/mainstream_test/_bulk").
+      with(:body => post_no_topics_to_elasticsearch).
+      to_return(:status => 200, :body => no_topics_elasticsearch_post_response)
+  end
+
+  def post_to_elasticsearch
```
Comment (rboulton) [inline test/support/index_document_test_helpers.rb]: This method name sounds like the method performs the action of posting.  Perhaps something like `elasticsearch_post_payload`?.  Similar for the next method.

File: test/support/index_document_test_helpers.rb [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+    "{\"index\":{\"_type\":\"edition\",\"_id\":\"/topic/animal-welfare/pets\"}}\n{\"popularity\":5.6085249579360626e-05,\"slug\":\"animal-welfare/pets\",\"topics\":[\"/ooo/aar\"],\"description\":\"Info about pets.\",\"format\":\"specialist_sector\",\"link\":\"/topic/animal-welfare/pets\",\"title\":\"Pets\",\"_type\":\"edition\",\"_id\":\"/topic/animal-welfare/pets\"}\n"
```
Comment (rboulton) [inline test/support/index_document_test_helpers.rb]: It's quite hard to read this; perhaps it would be clearer to write this as a nested hash datastructure and call .to_json on it to produce the sample body.

The topics should be being sent to elasticsearch in the `specialist_sectors` field, not the `topics` field.  (The `topics` field contains the things which whitehall calls topics internally, but which are now called "policy areas" in the UI.)

**Comment by bishboria** [comment]
Based on the number of comments made, I'll close this PR and reopen with different commits
